[eas-cli] include spawnProcess output in error message

kgc00 opened the pull request:
<!-- Thanks for contributing to _EAS CLI_! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->

# Checklist

- [ ] I've added an entry to [CHANGELOG.md](https://github.com/expo/eas-cli/blob/main/CHANGELOG.md#main) if necessary. You can comment this pull request with `/changelog-entry [breaking-change|new-feature|bug-fix|chore] [message]` and CHANGELOG.md will be updated automatically.

# Why

Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.

# How

How did you build this feature or fix this bug and why?

# Test Plan

Todo

ide commented:
Give a variant of Kim's idea a try. Instead of getting the stdout/stderr streams, I'd recommend passing process.stdout and stderr through to the child process. This way the child process's output will go straight to stdout + stderr.

Separately, depending on the thrown error, we could show a different helpful message. For instance, if expo-cli is not available at all (`error.code === 'ENOENT'`), we could show a nice message explaining that instead of showing an ENOENT error. It would still be a good idea to write `error.message` or `error.stack` when in debug mode though, if we have that pattern elsewhere in the CLI.

kgc00 commented:
> Give a variant of Kim's idea a try. Instead of getting the stdout/stderr streams, I'd recommend passing process.stdout and stderr through to the child process. This way the child process's output will go straight to stdout + stderr.

the end result of this is shown in the video above - we get the expo console output (which contains the error), then the error is repeated again below it. Is this what we want? or should we do some string parsing and only show the error?

there is some functionality already built into `expoCommandAsync` to pipe the output to the console:

https://user-images.githubusercontent.com/15132641/174352992-73cf86d3-d765-42c9-bd87-d556cc6042d4.mov

here, I did some string parsing to only show the error once:

https://user-images.githubusercontent.com/15132641/174353409-35cdfed8-d108-4493-a297-385febcf4e27.mov

Additionally:
- do we want to show verbose output for every call to `expoCommandAsync` or only the `expo export` command inside of `buildBundlesAsync`? 
- do we want the verbose behavior to be the default and allow for a new flag: `--silent` to suppress it?

ide commented:
I would expect stdout/stderr to be streamed and displayed only once. There's no parsing necessary.

Modifying every caller of expoCommandAsync is out of scope. I would evaluate that separately.

In this case I think streaming the exporter/bundler output is reasonable to do by default.

kgc00 commented:
> I would expect stdout/stderr to be streamed and displayed only once. There's no parsing necessary.

In that case I'll just flip the silent switch:

https://user-images.githubusercontent.com/15132641/174395905-c0b02573-30e3-4521-9371-aa97ca741923.mov

ide commented:
The latest video clip shows the gist of what we want but the spinner text showing up 4x looks sloppy. We either need to make `expoCommandAsync` render the text like a TTY or pass through stdio FDs directly.

kgc00 commented:
> The latest video clip shows the gist of what we want but the spinner text showing up 4x looks sloppy. We either need to make `expoCommandAsync` render the text like a TTY or pass through stdio FDs directly.

I'll put this back in draft and work on that. sorry for the back and forth

wkozyra95 commented:
> The latest video clip shows the gist of what we want but the spinner text showing up 4x looks sloppy. We either need to make expoCommandAsync render the text like a TTY or pass through stdio FDs directly.

if we pass output directly then we would lose that `[expo-cli]` prefix, I think it's important for users to see that an external program is called there.
To solve the spinner issue we could run that in non-interactive mode, previously command output was not showing any output, so I assume no interactive actions are required there?

As a side note, do we always need to display those logs? Wouldn't be better to print them only if there is an error(or debug mode)?

dsokal commented:
> As a side note, do we always need to display those logs? Wouldn't be better to print them only if there is an error(or debug mode)?

Agreed with Wojtek. I'd revert the change (so we only show logs when `EXPO_DEBUG` is set) but we can catch the error and print the logs only if sth went wrong.

ide commented:
The logs show the bundler's progress. Like most other scenarios where a bundler runs, we probably want to show progress all of the time, whether or not there's an error. When there is an error, we always want to show it without requiring a debug flag.

I don't have a strong opinion on whether the interactive spinner or the [expo-cli] prefix is more important, or if we should spend time implementing both by reimplementing a custom TTY stream. We could also print a line saying, "Building the update with Expo CLI:" followed by the Expo CLI output so that the terminal output doesn't have `[expo-cli]` prepended to  each line.

wkozyra95 commented:
> I believe it needs to be run in interactive mode

I'm not saying it's not, but that code from the screenshot does not really affect it. With the current implementation, there was no output so user would not see any prompts either way, `inherit` is set there most likely for other places where `expoCommandAsync` is called

kgc00 commented:
@wkozyra95 just following up again on 
> To solve the spinner issue we could run that in non-interactive mode

Indeed, this was the solution! Great suggestion 😄 

I'll be taking this back out of draft with amended behavior based on ^^

kgc00 commented:
updated failure recording:

https://user-images.githubusercontent.com/15132641/174918720-d8bde676-39f5-4c28-8875-8d1c795bee69.mov

updated success recording:

https://user-images.githubusercontent.com/15132641/174918858-d4067cf6-9d4f-41d9-a65c-2c82af7f6033.mov

Just showing the easiest-to-implement solution above. If we want differing behavior as talked about above, I can get on that. (for example, clearing out successful bundle cli output in non-error scenarios)

wkozyra95 commented on packages/eas-cli/src/project/publish.ts:
```diff
@@ -160,8 +159,8 @@ export async function buildBundlesAsync({
-    ['export', '--output-dir', inputDir, '--experimental-bundle'],
-    { silent: !Log.isDebug }
+    ['export', '--output-dir', inputDir, '--experimental-bundle', '--non-interactive'],
+    { silent: false }
```
```suggestion
```
false is default

dsokal reviewed (approved):
Add a changelog entry
